theano/gof/wrapper.py

```diff
@@ -0,0 +1,449 @@
+        return 'Wrapper<%s>' % ', '.join([('%s:%s' % (self.fields[i], self.types[i])) for i in range(self.length)])
+
+    def __eq__(self, other):
+        # To be checked.
+        return (type(self) == type(other) and self.fields == other.fields and self.types == other.types)
+
+    def __hash__(self):
+        return hash((type(self),) + self.fields + self.types)
+
+    def generate_struct_name(self):
+        # This method tries to generate an unique name for the current instance.
+        # This name is intended to be used as struct name in C code and as constant
+        # definition to check if a similar Wrapper has already been created
+        # (see c_support_code() below).
+        fields_string = ','.join(self.fields).encode('utf-8')
+        types_string = ','.join(str(t) for t in self.types).encode('utf-8')
+        fields_hex = hashlib.md5(fields_string).hexdigest()
+        types_hex = hashlib.md5(types_string).hexdigest()
+        return '_wrapper_%s_%s' % (fields_hex, types_hex)
+
+    def wrap_data(self, data, strict, allow_downcast):
+        # Try to wrap data. Raise an exception if data does not respect the Wrapper's contract.
+        wrap_instance = dict()
+        for i in range(self.length):
+            wrap_instance[self.fields[i]] = self.types[i].filter(getattr(data, self.fields[i]), strict, allow_downcast)
+        return data if strict else Wrap(self, **wrap_instance)
+
+    # Returns a wrapped object with expected attributes or (in strict mode) checks that data has expected attributes.
+    def filter(self, data, strict=False, allow_downcast=None):
+        if strict and not isinstance(data, Wrap):
+            raise TypeError('%s: strict mode: data should be an instance of Wrap.' % self)
+        return self.wrap_data(data, strict, allow_downcast)
+
+    def values_eq(self, a, b):
+        # We check that a and b have expected attributes and strict values.
+        a = self.filter(a, strict=True)
+        b = self.filter(b, strict=True)
+        # Then we compare.
+        return all(self.types[i].values_eq(getattr(a, self.fields[i]), getattr(b, self.fields[i]))
+                   for i in range(self.length))
+
+    def values_eq_approx(self, a, b):
+        # We check, wrap and round a and b if necessary.
+        a = self.filter(a, strict=False, allow_downcast=True)
+        b = self.filter(b, strict=False, allow_downcast=True)
+        # Then we compare.
+        return all(self.types[i].values_eq_approx(getattr(a, self.fields[i]), getattr(b, self.fields[i]))
+                   for i in range(self.length))
+
+    def c_compile_args(self, c_compiler):
+        c_compile_args_list = []
+        for _type in self.types:
+            try:
+                try:
+                    c_compile_args_list.extend(_type.c_compile_args(c_compiler))
+                except TypeError:
+                    c_compile_args_list.extend(_type.c_compile_args())
+            except MethodNotDefined:
+                pass
```

Comment:
This code makes me want to find a better way to do this.  But there are currently none.